azurelinuxagent/common/event.py

```diff
@@ -342,12 +335,36 @@ def add_metric(self, category, counter, instance, value, log_event=False):
         event.parameters.append(TelemetryEventParam('Instance', instance))
         event.parameters.append(TelemetryEventParam('Value', value))
 
+        self._add_default_parameters_to_event(event)
         data = get_properties(event)
         try:
             self.save_event(json.dumps(data))
         except EventError as e:
             logger.error("{0}", e)
 
+    @staticmethod
+    def _add_default_parameters_to_event(event):
+        # We write the GAVersion here rather than add it in azurelinuxagent.ga.monitor.MonitorHandler.add_sysinfo
+        # as there could be a possibility of events being sent with newer version of the agent, rather than the agent
+        # version generating the event.
+        # Old behavior example: V1 writes the event on the disk and finds an update immediately, and updates. Now the
+        # new monitor thread would pick up the events from the disk and send it with the CURRENT_AGENT, which would have
+        # newer version of the agent. This causes confusion.
+        event.parameters.append(TelemetryEventParam("GAVersion", CURRENT_AGENT))
+
+        # ContainerId can change due to live migration and we want to preserve the container Id of the container writing
+        # the event, rather than sending the event.
+        event.parameters.append(TelemetryEventParam('ContainerId', get_container_id_from_env()))
+
+        # This is used as the actual time of event generation.
+        event.parameters.append(TelemetryEventParam('OpcodeName', datetime.utcnow().__str__()))
+
+        # Other commong parameters added to the events
+        event.parameters.append(TelemetryEventParam('EventTid', threading.current_thread().ident))
+        event.parameters.append(TelemetryEventParam('EventPid', os.getpid()))
+        event.parameters.append(TelemetryEventParam("TaskName", threading.current_thread().getName()))
```

Comment:
I think we should reserve the TaskName for something else. We already have the identity of the thread in EventTid

Reply:
I think there's value in sending the thread name too, since only seeing a PID in Kusto doesn't tell us much. We can't map a PID to a thread name based only on the PID, right?

Reply:
There are only 2 threads with names: the daemon and the handler, so this doesn't add a lot of value. From the messages we can easily figure out daemon vs handler.

If preserving the names is important an alternative could be to use EventTid for that.

TaskName comes from the WIndows agent, where os events have this property. 

We could consider using it later for actual tasks in the agent ("download extension", "update goal state", "collect cgroup data", etc)

Reply:
BTW, the Windows Agent uses TPL's tasks, which are different than threads, and populate this field with the task name, so for them this does add value.
If we do not have a good use for this field, I'd leave it empty for future use. One use is what I mentioned above, but it could be something totally different. For example, this PR uses the opcode for the actual timestamp of the event. We may need another field in the future and having this one available would help.

Reply:
> There are only 2 threads with names: the daemon and the handler, so this doesn't add a lot of value. From the messages we can easily figure out daemon vs handler.

There are several thread names being set in this PR: MonitorHandler, ExtHandler, EnvHandler.

> If preserving the names is important an alternative could be to use EventTid for that.

If we're choosing between discarding the thread name completely or using EventTid for the thread name instead of the thread id, I vote for keeping the thread name and using EventTid.

Reply:
> using EventTid for the thread name instead of the thread id, I vote for keeping the thread name and using EventTid.

me too

Reply:
if doing that you may want to add an extra check to populate use the thread id if the name is not set

Reply:
>We can't map a PID to a thread name based only on the PID, right?

I haven't investigated that route.

Reply:
Yes, right now EventTid is not very useful, but as we add more events it'll be good to know which thread they are coming from.
As far as this PR, I'm ok with leaving it empty, use numeric thread id, or use combo of name/numeric id.

Reply:
Okay. We can take a new action item once we start seeing the usage of the new telemetry and usefulness of the new fields.